### ISSUE TITLE
integrating with other browsers

### ISSUE THREAD

**Issue by joeblew99** [knq/chromedp#17]
SO far in chrmeP there is not much support for other browsers yet, like Edge etc.

Murlok might be a good companion to help bridge this.
https://github.com/murlokswarm

There are drivers for webview based browsers. SO far OSX and Windows (super beta).
But Android and IOS are planned.

Murlok is golang based and concerned with allowing golang programmers to run HTML based apps easily with in webviews on all Desktops and Mobiles.

Would be curious what you think

**Comment by kenshaw** [comment]
I have done quite a bit of testing, actually with Safari, and with Microsoft Edge (through the Edge Diagnostics Adapter). While neither are 100% working right now (Edge has the biggest problems), I am planning on rolling out the necessary modifications/changes to enable them to work "out of the box" in the same way that Chrome is now. Also, at least for the browsers sharing WebKit ancestry, I've made a point to try to write the supplied Actions as backwards compatible. There are some gotchas/problems with how the debugging protocol has changed over time, that do make some APIs not backward compatible, but for the most part those are small-ish in nature and could easily be worked around when needed.

With some modifications to the generated API, however, the code does work now with Safari and Edge -- at least to some degree. The bigger issue for building out compatibility is that Google literally changes the debugging protocol on at least a weekly basis, and I haven't really decided yet what the best way to handle this is, and still be able to offer versions of the generated API that are compatible with older versions of Chrome, or the other browsers. Because I've chosen to make the API type safe, and because the protocol contains "hundreds" of API calls, types, etc. the code base would quickly become unreasonably large (as well as the built binaries!) if we tried to offer out of the box support for every browser.

At this juncture, I'm thinking that "out of the box" compatibility with Chrome's source tree, along with having the higher level Action code support the various, inevitable workarounds for variations between browsers/debugging protocol versions, allowing users to regenerate the API easily (they kind of can already) for other browser versions is the right "path". This would allow those who absolutely need to use a common code base for testing all those browsers to then supply / generate multiple versions of the generated API for all the versions they need to support.

I appreciate the link to Murlok, but I don't really see how it might be able to fix the issues with the generated API.
